src/ucx_uct_plugin.c

```diff
@@ -0,0 +1,1525 @@
+  return ncclSuccess;
+}
+
+static ncclResult_t nccl_uct_irecv(void *recv_comm, int n, void **data,
+                                   int *sizes, int *tags, void **mhandles,
+                                   void **request) {
+  nccl_uct_comm_t *comm      = recv_comm;
+  nccl_uct_memh_t **uct_memh = (nccl_uct_memh_t**)mhandles;
+  nccl_uct_rdesc_t *rdesc;
+  ucs_status_t status;
+
+  assert(n <= NCCL_UCX_UCT_MAX_RECVS);
+
+  rdesc = nccl_uct_comm_rdesc_get(comm);
+  if (rdesc == NULL) {
+    return ncclInternalError;
+  }
+
+  nccl_uct_rdesc_set(rdesc, comm->rdesc_id++, n, data, sizes, tags, uct_memh);
+
+  status = uct_ep_am_short(comm->uct_ep->ep, NCCL_UCT_AM_RTR,
+                           (uint64_t)comm->remote.comm, &rdesc->desc,
+                           nccl_uct_rdesc_size(n));
+  if (status != UCS_OK) {
+    nccl_uct_comm_rdesc_put(rdesc);
+    *request = NULL;
+  } else {
+    /* Wait for receiving ATP */
+    *request = nccl_uct_rdesc_get_req(rdesc, 0, NCCL_UCT_REQ_IRECV);
+  }
+
+  return ncclSuccess;
+}
+
+static ncclResult_t nccl_uct_iflush(void *recv_comm, int n, void **data,
+                                    int *sizes, void **mhandle,
+                                    void **request) {
+  int last                   = -1;
+  nccl_uct_comm_t *comm      = recv_comm;
+  nccl_uct_memh_t **uct_memh = (nccl_uct_memh_t**)mhandle;
+  nccl_uct_rdesc_t *rdesc;
+  nccl_uct_req_t *req;
+  ucs_status_t status;
+  uct_iov_t iov;
+  int i;
+
+  if (comm->gpu_flush.enabled) {
+    for (i = 0; i < n; i++) {
+      if (sizes[i]) {
+        last = i;
+      }
+    }
+  }
+
+  if (last == -1) {
+    return ncclSuccess;
+  }
+
+  rdesc = nccl_uct_comm_rdesc_get(comm);
+  if (rdesc == NULL) {
+    return ncclInternalError;
+  }
+
+  nccl_uct_rdesc_set(rdesc, ~0, 0, NULL, NULL, NULL, NULL);
+  /* Wait for local GET completion */
+  req = nccl_uct_rdesc_get_req(rdesc, 0, NCCL_UCT_REQ_IFLUSH);
+
+  iov.buffer = comm->gpu_flush.mem;
+  iov.length = comm->uct_iface->min_get_zcopy;
+  iov.memh   = comm->gpu_flush.memh;
+  iov.stride = 0;
+  iov.count  = 1;
+
+  status = uct_ep_get_zcopy(comm->gpu_flush.uct_ep->ep, &iov, 1,
+                            (uint64_t)data[last], uct_memh[last]->bundle.rkey,
+                            &req->completion);
+  if (status == UCS_OK) {
+    *request = NULL;
+    nccl_uct_comm_rdesc_put(rdesc);
+  } else if (status == UCS_INPROGRESS) {
+    *request = req;
+  } else {
+    WARN("Failed to flush local ep comm=%p status=%d", comm, status);
+    return ncclInternalError;
+  }
+
+  return ncclSuccess;
+}
+
+static ncclResult_t nccl_uct_test(void *request, int *done, int *sizes) {
+  nccl_uct_req_t *req     = request;
+  nccl_uct_rdesc_t *rdesc = req->rdesc;
+  nccl_uct_comm_t *comm   = rdesc->comm;
+
+  uct_worker_progress(comm->uct_worker->worker);
+
+  *done = 0;
+
+  if (rdesc->send_atp == 1) {
+    /* Slowpath */
+    nccl_uct_send_atp(comm, rdesc);
+
+    if (rdesc->send_atp && rdesc->nccl_usage == 1) {
```

Comment:
why need to check also `nccl_usage`?

Reply:
there can be multiple nccl send request (`->isend()`) for a given `rdesc`. if nccl_usage == 1, we are progressing the last non-released nccl send request. we need to keep it in order to trigger retry until send atp succeeds.

but from nccl point of view, sends with complete.count == 0 can be marked as completed, even before atp has started.